integration/loki_micro_services_test.go

```diff
@@ -708,6 +706,128 @@ func TestQueryTSDB_WithCachedPostings(t *testing.T) {
+		)
+		tQueryScheduler = clu.AddComponent(
+			"query-scheduler",
+			"-target=query-scheduler",
+			"-query-scheduler.use-scheduler-ring=false",
+			"-boltdb.shipper.index-gateway-client.server-address="+tIndexGateway.GRPCURL(),
+		)
+	)
+	require.NoError(t, clu.Run())
+
+	// finally, run the query-frontend and querier.
+	var (
+		tQueryFrontend = clu.AddComponent(
+			"query-frontend",
+			"-target=query-frontend",
+			"-frontend.scheduler-address="+tQueryScheduler.GRPCURL(),
+			"-boltdb.shipper.index-gateway-client.server-address="+tIndexGateway.GRPCURL(),
+			"-common.compactor-address="+tCompactor.HTTPURL(),
+			"-querier.per-request-limits-enabled=true",
+			"-frontend.required-query-response-format=protobuf",
+		)
+		_ = clu.AddComponent(
+			"querier",
+			"-target=querier",
+			"-querier.scheduler-address="+tQueryScheduler.GRPCURL(),
+			"-boltdb.shipper.index-gateway-client.server-address="+tIndexGateway.GRPCURL(),
+			"-common.compactor-address="+tCompactor.HTTPURL(),
+		)
+	)
+	require.NoError(t, clu.Run())
+
+	tenantID := randStringRunes()
+
+	now := time.Now()
+	cliDistributor := client.New(tenantID, "", tDistributor.HTTPURL())
+	cliDistributor.Now = now
+	cliIngester := client.New(tenantID, "", tIngester.HTTPURL())
+	cliIngester.Now = now
+	cliQueryFrontend := client.New(tenantID, "", tQueryFrontend.HTTPURL())
+	cliQueryFrontend.Now = now
+
+	t.Run("ingest-logs", func(t *testing.T) {
+		// ingest some log lines
+		require.NoError(t, cliDistributor.PushOTLPLogLine("lineA", now.Add(-45*time.Minute), map[string]any{"trace_id": 1, "user_id": "2"}))
+		require.NoError(t, cliDistributor.PushOTLPLogLine("lineB", now.Add(-45*time.Minute), nil))
+
+		require.NoError(t, cliDistributor.PushOTLPLogLine("lineC", now, map[string]any{"order.ids": []any{5, 6}}))
+		require.NoError(t, cliDistributor.PushOTLPLogLine("lineD", now, nil))
+	})
+
+	t.Run("query", func(t *testing.T) {
+		resp, err := cliQueryFrontend.RunRangeQuery(context.Background(), `{service_name="varlog"}`)
+		require.NoError(t, err)
+		assert.Equal(t, "streams", resp.Data.ResultType)
+
+		for i, stream := range resp.Data.Stream {
+			switch i {
+			case 0:
+				require.Len(t, stream.Values, 2)
+				require.Equal(t, "lineD", stream.Values[0][1])
+				require.Equal(t, "lineB", stream.Values[1][1])
+				require.Equal(t, map[string]string{
+					"service_name": "varlog",
+				}, stream.Stream)
+			case 1:
+				require.Len(t, stream.Values, 1)
+				require.Equal(t, "lineA", stream.Values[0][1])
+				require.Equal(t, map[string]string{
+					"service_name": "varlog",
+					"trace_id":     "1",
+					"user_id":      "2",
+				}, stream.Stream)
+			case 2:
+				require.Len(t, stream.Values, 1)
+				require.Equal(t, "lineC", stream.Values[0][1])
+				require.Equal(t, map[string]string{
+					"service_name": "varlog",
+					"order_ids":    "[5,6]",
+				}, stream.Stream)
+			default:
+				t.Errorf("unexpected case %d", i)
+			}
+		}
```

Comment:
these assertions do not make sure that there are exactly 3 log lines in the response. If the response has only 1 line it will assert the first line and will mark the test as passed